### ISSUE TITLE
WPF Localization Part 2

### ISSUE THREAD

**Issue by Ryochan7** [Ryochan7/DS4Windows#883]
So far, getting the initial localization hooks configured for the WPF build has been a pain in the butt. The official WPF Localize Extension documentation covers quite a bit of the process but it leaves out some critical details that I had to find resolutions for. On my local copy, Design and Runtime localization now work.

On the Design side, the biggest problem is that the system only works while targeting platform x86; targeting x64 will just show blank strings in the Design view with no way to provide a fallback. At Runtime, the extension has a couple of extra calls that need to be performed outside the normal culture change process within .NET and the official documentation does not cover that; had to find the resolution in other source code and tutorials. One other problem on the Design side is regarding resource caching. Any change in the Resource strings requires explicitly killing the XAML Designer and reloading it in order for the changes to be visible in the Designer.

The Language selector combobox, used in the Settings tab, has already been converted and it does detect relevant language assemblies. The setting is not hooked into the application yet so changing the chosen language pack does nothing in the GUI.

Other than English, I will attempt to convert a bit of the current Japanese localization strings to the WPF build so there will be another language assembly to test against. All other current language packs will be left to other people to provide translations for. The current keys used in the main Resources.resx file will still be used but the form specific resx files will have to be remade with new keys for identifiers. I need to decide on a standard and get more controls localized.

**Comment by Ryochan7** [comment]
It looks like the lack of translations in the Design view in x64 mode might actually be a limitation of Visual Studio itself. It does not seem like the **Run Project Code** Design option is avalable while targeting platform x64. Funny enough, disabling the option while in x86 mode and switching back to x64 mode at least allows the WPF Localize Extension to show the key name as a string rather than rendering nothing.

**Comment by Ryochan7** [comment]
Continuing from the previous post, using Blend alongside Visual Studio might be useful while developing translations. That way Blend can be configured for the x86 target so the GUI can be rendered localized while Visual Studio can be used to edit other backend code in x64 mode. The following page shows an example of using both tools to handle design and coding when targeting different platforms.

https://dzone.com/articles/simple-workaround-%E2%80%9Cdesign-view

I am kind of lost at the moment regarding the structure of the RESX files to use. Initially, I was thinking about attempting to duplicate the Window Forms way of doing things with a separate RESX file being associated with each form. Other WPF projects that I have looked at use one main RESX file throughout the project with the culture specific files alongside it.

**Comment by Ryochan7** [comment]
I think I will go with option 2. It might be harder for development in the long run to keep the neutral translatable strings contained in one file. Although, it will likely be easier for translators for the strings to be contained in one file rather than scattered among many files like how Windows Forms handles localization; for reference, the BAML method would have used one CSV file if that approach would have been used. Another option that opens up is maybe using online translation suites to manage collaborating on translation work.

https://weblate.org/

**Comment by Ryochan7** [comment]
Now actual localization work can begin. There is a starting structure in place and I added a couple of translation strings so far. Things will likely change but at least the localization extension can now get actively used in the program.

**Comment by Ryochan7** [comment]
The current Japanese translation has been fully ported over to the WPF build. It is too bad that the Japanese translation was very incomplete to begin with. I had to cross reference some kanji on Jisho while porting. I will try to get some more of the GUI translated with my limited Nihongo knowledge and using Jisho.org for reference.

**Comment by Ryochan7** [comment]
Got a bit more done with this. The final hook is now in place and the WPF build will now use the assigned language pack from the Settings tab. A new WPF beta build will probably be uploaded later today.

**Comment by Ryochan7** [comment]
Here is an WPF beta build with localization supported added.

https://drive.google.com/file/d/1B4Q4fbF31OgBaeZayB5yoSs3SIB-5x6F/view?usp=sharing

**Comment by Ryusennin** [comment]
I'm not familiar with .NET resource files, but I tried my hand at a French translation of the Strings.resx file with ResX Editor. I checked all exposed strings with DS4Windows opened side by side, but it appears a good number of options and buttons are not in the translation file (like "Hide DS4 Controller"). I suppose the localization is a work in progress, but is this expected or did I miss another file?
